### PR TITLE
feat: implement rollup module

### PR DESCRIPTION
**Pull request by shiki-tak**
## Description

Implement the x/rollup module to prepare for the rollup to start.

**The demo runs with the script below.**
https://github.com/Finschia/finschia-sdk/pull/1034/files#diff-996183dd2e1b9a5dcf80b6875e7e0631b9ebf854e79a1214086b7facfa5d1af4R1

1. Create Rollup
- Register the rollup you want to start. Currently only allows rollup:sequencer=1:1.
  - https://github.com/Finschia/finschia-sdk/pull/1034/files#diff-2154285ebe7da9a690d97756b64257405cbbd048c12f6919aea452bd9ca62aacR42

2. Register Sequencer
- Register the sequencer for the created rollup.
  - https://github.com/Finschia/finschia-sdk/pull/1034/files#diff-2154285ebe7da9a690d97756b64257405cbbd048c12f6919aea452bd9ca62aacR96
- At this time, specify the amount flag and deposit to become a sequencer. The minimum deposit amount at this time is defined below. We will make this changeable later.
  - https://github.com/Finschia/finschia-sdk/pull/1034/files#diff-38e1147424e4fdc69d3a7f9961aca038f4ac24be3406a4a2e5c467f21776e6daR13

3. Slash
- If the sequencer doesn't work properly, slash the deposit. It is not assumed that this is done by command execution of account, and cli does not implement it.
  - https://github.com/Finschia/finschia-sdk/pull/1034/files#diff-2154285ebe7da9a690d97756b64257405cbbd048c12f6919aea452bd9ca62aacR143

4. Deposit/Withdraw
- It is also possible to deposit/withdraw the sequencer later.
  - https://github.com/Finschia/finschia-sdk/pull/1034/files#diff-2154285ebe7da9a690d97756b64257405cbbd048c12f6919aea452bd9ca62aacR100
  - https://github.com/Finschia/finschia-sdk/pull/1034/files#diff-2154285ebe7da9a690d97756b64257405cbbd048c12f6919aea452bd9ca62aacR176

## Checklist:

- [ ] I followed the [contributing guidelines](https://github.com/Finschia/finschia-sdk/blob/main/CONTRIBUTING.md) and [code of conduct](https://github.com/Finschia/finschia-sdk/blob/main/CODE_OF_CONDUCT.md).
- [ ] I have added a relevant changelog to `CHANGELOG.md`
- [ ] I have added tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have updated API documentation `client/docs/swagger-ui/swagger.yaml`

### CODE REVIEW

File: x/or/rollup/keeper/msg_server.go [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+		SequencerAddress: msg.SequencerAddress,
+		Value:            msg.Value,
+	}
+
+	k.SetDeposit(ctx, deposit)
+
+	return &types.MsgDepositResponse{}, nil
+}
+
+func (k msgServer) Slash(ctx sdk.Context, rollupName string, sequencerAddress string, value sdk.Coin) error {
```
Comment (ulbqb) [inline x/or/rollup/keeper/msg_server.go:142]: I think this function is used by only inner modules. Would you like to move this function to sequencer.go?

File: proto/finschia/or/rollup/v1/genesis.proto [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+import "gogoproto/gogo.proto";
+
+import "finschia/or/rollup/v1/rollup.proto";
+import "finschia/or/rollup/v1/params.proto";
+
+option go_package = "github.com/Finschia/finschia-sdk/x/or/rollup/types";
+
+message GenesisState {
+    repeated Rollup rollupList = 1 [(gogoproto.nullable) = false];
```
Comment (ulbqb) [inline proto/finschia/or/rollup/v1/genesis.proto:12]: Please rewrite all proto fields in snake_case.

**Review by ulbqb — CHANGES_REQUESTED** [review]

File: x/or/demo/prepare_chain.sh [hunk 3]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
```
[the remark this answers is not in the document]
Comment (ulbqb) [reply x/or/demo/prepare_chain.sh:1]: I would like you to enable to use various addresses to use my case.

```bash
#!/bin/bash

# Remove data
pkill simd
rm -rf ~/.simapp

# Prepare chain
TEST_MNEMONIC="mind flame tobacco sense move hammer drift crime ring globe art gaze cinnamon helmet cruise special produce notable negative wait path scrap recall have"
BASE_DIR=~/.simapp
CHAIN_ID=sim

simd init validator --home $BASE_DIR --chain-id $CHAIN_ID > /dev/null 2>&1
simd keys add validator --keyring-backend=test --home $BASE_DIR --recover --account=0 <<< ${TEST_MNEMONIC} > /dev/null 2>&1
simd keys add sequencer --keyring-backend=test --home $BASE_DIR --recover --account=1 <<< ${TEST_MNEMONIC} > /dev/null 2>&1
simd keys add challenger --keyring-backend=test --home $BASE_DIR --recover --account=2 <<< ${TEST_MNEMONIC} > /dev/null 2>&1
simd add-genesis-account $(simd --home $BASE_DIR keys show validator -a --keyring-backend=test) 100000000000stake,100000000000tcony --home $BASE_DIR > /dev/null 2>&1
simd add-genesis-account $(simd --home $BASE_DIR keys show sequencer -a --keyring-backend=test) 100000000000stake,100000000000tcony --home $BASE_DIR > /dev/null 2>&1
simd add-genesis-account $(simd --home $BASE_DIR keys show challenger -a --keyring-backend=test) 100000000000stake,100000000000tcony --home $BASE_DIR > /dev/null 2>&1
simd gentx validator 10000000000stake --keyring-backend=test --home $BASE_DIR --chain-id=$CHAIN_ID > /dev/null 2>&1
simd collect-gentxs --home $BASE_DIR > /dev/null 2>&1

# Run chain
simd start --home $BASE_DIR > $BASE_DIR/$CHAIN_ID.log 2>&1 &
```

File: x/or/rollup/client/cli/tx.go [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+			}
+
+			return tx.GenerateOrBroadcastTxCLI(clientCtx, cmd.Flags(), msg)
+		},
+	}
+
+	cmd.Flags().AddFlagSet(FlagSetAmount())
+	flags.AddTxFlagsToCmd(cmd)
+
+	_ = cmd.MarkFlagRequired(FlagAmount)
```
Comment (ulbqb) [inline x/or/rollup/client/cli/tx.go:129]: Why is amount is the flag? How about add it to arguments?

File: proto/finschia/or/rollup/v1/rollup.proto [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+import "cosmos/base/v1beta1/coin.proto";
+import "gogoproto/gogo.proto";
+
+option go_package = "github.com/Finschia/finschia-sdk/x/or/rollup/types";
+
+message Rollup {
+  string rollupName = 1;
+  string creator     = 2;
+  Sequencers permissionedAddresses = 3 [(gogoproto.nullable) = false];
+  uint64 maxSequencers = 4;
```
[the remark this answers is not in the document]
Comment (shiki-tak) [reply proto/finschia/or/rollup/v1/rollup.proto:15]: If multiple sequencers can be registered for one rollup, they will be used.
Now it is fixed at 1.

https://github.com/Finschia/finschia-sdk/pull/1034/files#diff-29fc36437b7c2481fedf4fcd616f2f1aa9b47a35ed9a0cc07341f45a8feb266aR68

[the remark this answers is not in the document]
Comment (Mdaiki0730) [reply proto/finschia/or/rollup/v1/rollup.proto:15]: I think the value specified in MsgCreateRollup will be this value, but is it okay to set this value in request?
I think it is better to be decided inside the rollup module rather than being decided by the user issuing the request.
https://github.com/Finschia/finschia-sdk/pull/1034/files/dcd6d654e7eb97df207c822c20abb4dceafacc32#diff-9dfff786e5d6f6a569ad5afac902871268457bb8ecd3b9c99bc60be8d771bf89R24

File: x/or/rollup/keeper/msg_server.go (hunk 1 above)
Comment (shiki-tak) [reply x/or/rollup/keeper/msg_server.go:142]: I want to implement only Set*** and Get*** in sequencer.go, so move to keeper.go.

File: x/or/rollup/client/cli/query.go [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+			return clientCtx.PrintProto(res)
+		},
+	}
+
+	flags.AddQueryFlagsToCmd(cmd)
+
+	return cmd
+}
+
+// func NewQueryCmdShowSequencer() *cobra.Command {
```
[the remark this answers is not in the document]
Comment (shiki-tak) [reply x/or/rollup/client/cli/query.go:62]: I had already implemented it like this:
https://github.com/Finschia/finschia-sdk/pull/1034/files/dcd6d654e7eb97df207c822c20abb4dceafacc32#diff-9d2b7f072618cfeca68ff14de9e8060e3005151ba0b311d46da87d7dda0ac326R130

Remove comment out.

File: x/or/rollup/client/cli/tx.go (hunk 4 above)
Comment (shiki-tak) [reply x/or/rollup/client/cli/tx.go:123]: That was totally my preference, but I will modify the parameters to match the bank module etc.

**Comment by shiki-tak** [comment]
@ulbqb @Mdaiki0730 PTAL

**Review by ulbqb — APPROVED** [review]

**Review by Mdaiki0730 — APPROVED** [review]
LGTM
